nixos/modules/services/x11/imwheel.nix

```diff
@@ -0,0 +1,69 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.xserver.imwheel;
+in
+  {
+    options = {
+      services.xserver.imwheel = {
+        enable = mkEnableOption "IMWheel service";
+
+        options = mkOption {
+          type = types.listOf types.str;
+          default = ["--detach" "--buttons 45" "--kill"];
```

Comment:
I'd do some changes regarding this:
- If users set this to e.g. `["--buttons 45"]` they would override the `--detach` option (which actually should be called `--no-detach` or so anyways), so the service would fork off, which is not as nicely handled anymore then. So I suggest to add `--detach` unconditionally in the `ExecStart` to prevent this. I don't see any reason anybody would want to change that anyways.
- The name `options` is already an overloaded term from the module system, and with above change, `extraOptions` is more fitting, which is also the standard naming for such options.
- Unless I'm missing something, I don't see a reason `--buttons 45` should be passed by default. Same with `--kill` which doesn't make any sense here, or does it? So I'd just make `default = []` but put some examples in `example = [ "--debug" "--buttons 45" ]`

Reply:
These defaults are based mostly on my experimentation because sadly running IMWheel without flags causes some issues

- `detach` is by default **true** so passing `--detach` toggles it so **it doesn't detach**. Which I believe makes more sense when it's ran as a service.
- `kill` meas that any already running IMWheel process will be killed when this one starts. I think this also make sense because otherwise result can be unpredictable
- `buttons 45` is passed because without this IMWheel breaks functionality of additional mouse buttons like back and forward (tested with Logitech and MAD Rat)

Using defaults make it still possible for user to easily overwrite those things. This is why there is a `pkill` https://github.com/NixOS/nixpkgs/pull/71052/files#diff-ea25b843a6681ed1b7711805ec769667R64 so in case `--detach` is not passed (and so process is detached) IMWheel is still killed when service stops. This is also related to this comment https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544270064 and my reaction https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544281927.

so TL;DR:

- `--detach` is to prevent detaching which imo makes more sense for a service
- `--buttons 45` is to avoid breaking addition mouse buttons with default settings (just enable)
- `--kill` is to avoid multiple IMWheel processes
- `pkill` is used in cases someone overwrides defaults but doesn't pass `--detach` or `-d`

Which this in mind do you think want to change these defaults @Infinisil ?

Reply:
sorry for ping @Infinisil I don't want to be too pushy about this. I'm happy to implement any changes, just want to make sure you still feel we want to do them considering the comment above.

Reply:
It seems regardless of any changes to the defaults, the rename from `options` to `extraOptions` should happen.

Reply:
Hm okay I can see the `buttons` default being reasonable, but `--detach` and maybe `--kill` too can imo be added like
```nix
{
  ExecStart = "${pkgs.imwheel}/bin/imwheel" + escapeShellArgs ([
    "--detach"
    "--kill"
  ] ++ cfg.extraOptions);
}
```

This means even when the user doesn't specify them in `extraOptions`, they still get added, and the expectation is that users won't add `--detach` in `extraOptions`. It should feel like "Extra options in addition to the ones NixOS adds on its own because they always make sense to use"

Reply:
Thanks @Infinisil for the feedback. I've incorporate this change. just few remarks:

- I've add space behind `bin/imwheel` because I don't think escapeShellArgs do that. Let me know if that's not right.
- There is still the pkill on `ExecStop`. With these additional changes it might make sense to incorporate even [change to Simple type](https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544270064) - depends of if we want to try to protect user from shooting him/herself in foot (multiple --detach arguments case) or we want to make assumption default is good and if someone overrides it by compination of extraOptions then it's his/her thing to deal with.